### PR TITLE
refactor: various stdlib fixes for algorithm w

### PR DESCRIPTION
**Pull request by jsternberg**
This contains various fixes to the stdlib to make the tests pass with
algorithm w. It updates some of the builtins with the correct type
signature. Some of them did not match the Go code.

It updates `map` to work with compiled functions.

It also updates an end to end test which relied on polymorphic literals.

### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: semantic/graph.go [hunk 1]
```diff
@@ -499,7 +499,7 @@ type ArrayExpression struct {
 
 	Elements []Expression
 
-	typ MonoType
+	Type MonoType
```
Comment (nathanielc) [inline semantic/graph.go]: Why the change to make it public? The expression interface has the `TypeOf` method?

File: stdlib/universe/filter_mixed_empty_test.flux [hunk 2]
```diff
@@ -59,7 +59,7 @@ t_filter_mixed_empty = (table=<-) =>
   table
   |> range(start: 2018-05-22T19:53:26Z)
   |> filter(fn: (r) => r._measurement == "diskio")
-  |> filter(fn: (r) => r["_value"] > 1000, onEmpty: "keep")
+  |> filter(fn: (r) => r["_value"] > 1000.0, onEmpty: "keep")
```
Comment (nathanielc) [inline stdlib/universe/filter_mixed_empty_test.flux]: This should have worked correct? I though we removed the constraints about equal types for now?

File: semantic/graph.go (hunk 1 above)
Comment (jsternberg) [reply semantic/graph.go]: I think `resolveValue` needs to be refactored, but it's in a different package and so I don't have access to modify it. I am not sure `resolveValue` is needed anymore and, in my opinion, it is a bit of a code smell that it converts a value into a semantic node.

File: stdlib/universe/filter_mixed_empty_test.flux (hunk 2 above)
Comment (jsternberg) [reply stdlib/universe/filter_mixed_empty_test.flux]: Oh if we did then I need to rebase. I'll look at this. If it's not done, I can remove that constraint.

File: compiler/runtime.go [hunk 3]
```diff
@@ -389,7 +389,7 @@ func (e *unaryEvaluator) Eval(ctx context.Context, scope Scope) (values.Value, e
 	if err != nil {
 		return nil, err
 	}
-	values.CheckKind(ret.Type().Nature(), e.t.Nature())
+	// values.CheckKind(ret.Type().Nature(), e.t.Nature())
```
Comment (ayang64) [inline compiler/runtime.go]: i don't think you meant to commit commented out code.

Comment (jsternberg) [reply compiler/runtime.go]: We're mostly removing these checks because they are more often wrong than correct. We may re-add them so I've been commenting them out whenever they cause a problem.

File: stdlib/universe/filter_mixed_empty_test.flux (hunk 2 above)
Comment (jsternberg) [reply stdlib/universe/filter_mixed_empty_test.flux]: Updated.

File: libflux/src/flux/semantic/nodes.rs [hunk 4]
```diff
@@ -976,36 +976,44 @@ impl BinaryExpr {
...
+                // Constraint::Equal(self.left.type_of().clone(), self.right.type_of().clone()),
                 Constraint::Equal(self.typ.clone(), MonoType::Bool),
                 Constraint::Kind(self.left.type_of().clone(), Kind::Equatable),
                 Constraint::Kind(self.left.type_of().clone(), Kind::Comparable),
+                Constraint::Kind(self.right.type_of().clone(), Kind::Equatable),
+                Constraint::Kind(self.right.type_of().clone(), Kind::Comparable),
             ]),
             ast::Operator::LessThanEqualOperator => Constraints::from(vec![
-                Constraint::Equal(self.left.type_of().clone(), self.right.type_of().clone()),
+                // Constraint::Equal(self.left.type_of().clone(), self.right.type_of().clone()),
```
Comment (nathanielc) [inline libflux/src/flux/semantic/nodes.rs]: Can you add a link to this issue as part of these comments? https://github.com/influxdata/flux/issues/2393 that way we have the context for why they are commented out for now.

File: semantic/graph.go [hunk 5]
```diff
@@ -934,7 +934,7 @@ func (l *BooleanLiteral) Copy() Node {
 	return nl
 }
 func (e *BooleanLiteral) TypeOf() MonoType {
-	return e.typ
+	return BasicBool
```
Comment (jlapacik) [inline semantic/graph.go]: I would prefer if `TypeOf` returned the inferred type. I'd like to know if we ever infer anything other than `bool`.

Comment (jsternberg) [reply semantic/graph.go]: In the future we can remove this. The problem is resolving functions currently replaces values with semantic nodes and so there's no type inference done. We can hopefully remove this in the future because I think you're right.

File: libflux/src/flux/semantic/nodes.rs [hunk 6]
```diff
@@ -976,36 +976,50 @@ impl BinaryExpr {
                 Constraint::Kind(self.typ.clone(), Kind::Divisible),
             ]),
             ast::Operator::GreaterThanOperator => Constraints::from(vec![
-                Constraint::Equal(self.left.type_of().clone(), self.right.type_of().clone()),
+                // https://github.com/influxdata/flux/issues/2393
+                // Constraint::Equal(self.left.type_of().clone(), self.right.type_of().clone()),
                 Constraint::Equal(self.typ.clone(), MonoType::Bool),
                 Constraint::Kind(self.left.type_of().clone(), Kind::Comparable),
+                Constraint::Kind(self.right.type_of().clone(), Kind::Comparable),
```
Comment (jlapacik) [inline libflux/src/flux/semantic/nodes.rs]: This change probably should have been pointed directly at master since it's not `use-algow` specific
